mempool/src/pool/tests/orphans.rs

```diff
@@ -0,0 +1,280 @@
+    let genesis_id = tf.genesis().get_id();
+
+    let tx0 = TransactionBuilder::new()
+        .add_input(
+            TxInput::new(OutPointSourceId::BlockReward(genesis_id.into()), 0),
+            empty_witness(&mut rng),
+        )
+        .add_output(TxOutput::Transfer(
+            OutputValue::Coin(amt0),
+            Destination::AnyoneCanSpend,
+        ))
+        .build();
+    let tx0_id = tx0.transaction().get_id();
+
+    let tx1 = TransactionBuilder::new()
+        .add_input(
+            TxInput::new(OutPointSourceId::Transaction(tx0_id), 0),
+            empty_witness(&mut rng),
+        )
+        .add_output(TxOutput::Transfer(
+            OutputValue::Coin(amt1),
+            Destination::AnyoneCanSpend,
+        ))
+        .build();
+    let tx1_id = tx1.transaction().get_id();
+
+    let mut mempool = setup_with_chainstate(tf.chainstate()).await;
+
+    // Add the second transaction first
+    let res = mempool.add_transaction(tx1);
+    assert_eq!(res, Ok(TxStatus::InOrphanPool));
+    assert!(mempool.contains_orphan_transaction(&tx1_id));
+    assert!(!mempool.contains_transaction(&tx1_id));
+
+    // Now add the first transaction
+    let res = mempool.add_transaction(tx0);
+    assert_eq!(res, Ok(TxStatus::InMempool));
+
+    // Now the second tx (which we submitted first) should be either rejected or in mempool
+    assert!(!mempool.contains_orphan_transaction(&tx1_id));
+    assert_eq!(mempool.contains_transaction(&tx1_id), expected_in_mempool);
+}
+
+// Below, each test case encodes a sequence of transaction insertions.
+//
+// Each element of the Vec contains:
+// 0. Transaction to be inserted (see the diagram below)
+// 1. Expected state of the main (non-orphan) mempool. One bit per transaction, 1 = transaction is
+//    expected in mempool, 0 = transaction is expected not to be in mempool. Note that the lowest
+//    bit (rightmost digit) corresponds to transaction 0.
+// 2. Expected state of the orphan pool. Same encoding as above.
+//
+// The transaction topology used by the test is as follows:
+//
+// (i = input, o = output, [n] = transaction number as referred to in the code)
+//
+//                                   +-------+
+//                               +---i  [1]  o---+
+//                   +-------+   |   +-------+   |   +-------+
+// genesis UTXO -----i       o---+               +---i       o
+//                   |  [0]  |                       |  [3]  |
+//                   |       o---+               +---i       |
+//                   +-------+   |   +-------+   |   +-------+
+//                               +---i  [2]  o---+
+//                                   +-------+
+//
+// This is a simple diamond transaction graph but it covers many interesting cases.
+//
+#[rstest]
+#[trace]
+#[case::topological_order(
+    Seed::from_entropy(),
+    vec![(0, 0b0001, 0b0000), (1, 0b0011, 0b0000), (2, 0b0111, 0b0000), (3, 0b1111, 0b0000)],
+)]
```

Comment:
I noticed you can name test cases in rstest (the identifier right after `case::`). Pretty useful for documentation and running individual cases from the command line.

Reply:
Nice! Good to know.